pyfarm/master/api/tasklogs.py

```diff
@@ -166,26 +168,38 @@ def post(self, job_id, task_id, attempt):
         path = realpath(join(LOGFILES_DIR, g.json["identifier"]))
         if not realpath(path).startswith(LOGFILES_DIR):
             return jsonify(error="Identifier is not acceptable"), BAD_REQUEST
-        task_log = TaskLog.query.filter_by(
-            identifier=g.json["identifier"]).first()
-        if not task_log:
-            task_log = TaskLog(**g.json)
-
-        association = TaskTaskLogAssociation.query.filter_by(
-            task=task, log=task_log, attempt=attempt).first()
-        if association:
-            return (jsonify(
-                log=task_log, attempt=attempt, task_id=task_id,
-                error="This log is already registered for this task"), CONFLICT)
-
-        association = TaskTaskLogAssociation()
-        association.task = task
-        association.log = task_log
-        association.attempt = attempt
-
-        db.session.add(association)
-        db.session.add(task_log)
-        db.session.commit()
+        attempts = 0
+        registered = False
+        while not registered:
+            try:
+                task_log = TaskLog.query.filter_by(
+                    identifier=g.json["identifier"]).first()
+                if not task_log:
+                    task_log = TaskLog(**g.json)
+
+                association = TaskTaskLogAssociation.query.filter_by(
+                    task=task, log=task_log, attempt=attempt).first()
+                if association:
+                    return (jsonify(
+                        log=task_log, attempt=attempt, task_id=task_id,
+                        error="This log is already registered for this task"),
+                    CONFLICT)
+
+                association = TaskTaskLogAssociation()
+                association.task = task
+                association.log = task_log
+                association.attempt = attempt
+
+                db.session.add(association)
+                db.session.add(task_log)
+                db.session.commit()
+                registered = True
+            except IntegrityError:
```

Comment:
> Hmm, I wonder if the `IntegrityError` instance has any useful attributes on it that could help us figure out why this happens?

It's a race condition.

The agent is registering the new tasklog on all the tasks in its assignment at once. There are several processes on the master running simulataneously, all checking whether this tasklog record already exists in the database. All of them get the result that, no, no such tasklog exists yet, so they all try to create it. But of course, only one of them can actually succeed, and the others will get a unique constraint violation.

> It might also make some sense to only wrap the code that's failing in the while loop here so we don't end up having to redo the queries for example.

No. When you do a database transaction rollback in order to do the same transaction again, you MUST throw away all data you got from the failed transaction. If you don't, you're not working on a consistent view of the database.